src/libmongoc/src/mongoc/mongoc-generation-map.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2021-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-generation-map-private.h"
+
+#include "utlist.h"
+
+typedef struct _gm_node_t {
+   bson_oid_t key;
+   uint32_t val;
+   struct _gm_node_t *next;
+} gm_node_t;
+
+static gm_node_t *
+gm_node_new ()
+{
+   return bson_malloc0 (sizeof (gm_node_t));
+}
+
+static void
+gm_node_destroy (gm_node_t *node)
+{
+   if (!node) {
+      return;
+   }
+   bson_free (node);
+}
+
+static gm_node_t *
+gm_node_copy (const gm_node_t *node)
+{
+   gm_node_t *node_copy = gm_node_new ();
```

Comment:
Is it ok that the result of this is not checked?
(Peeking at bson_oid_copy() below, it uses BSON_ASSERT() on both of its parameters; however, since the "key" member is being dereferenced, if that happens on a null pointer, I expect it will explode.)